File: app/src/androidTest/java/com/amyllykoski/asyncseq/MyServiceImplTest.java
package com.amyllykoski.asyncseq;

import android.support.test.espresso.core.internal.deps.guava.base.Predicate;

import com.amyllykoski.asyncseq.api.MyService;
import com.amyllykoski.asyncseq.api.RestCallback;
import com.amyllykoski.asyncseq.impl.MyServiceImpl;
import com.amyllykoski.asyncseq.model.Item;
import com.amyllykoski.asyncseq.util.L;
import com.amyllykoski.asyncseq.util.RandomString;
import com.google.gson.Gson;

import org.junit.Before;
import org.junit.Test;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ThreadLocalRandom;

import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;

import static junit.framework.Assert.assertTrue;
import static junit.framework.Assert.fail;

public class MyServiceImplTest {
  private static final String TAG = MyServiceImplTest.class.getSimpleName();
  private MockWebServer server;

  @Before
  public void setUp() throws Exception {
    server = new MockWebServer();
    server.start();
  }

  @Test
  public void getItemsFromMockServer() throws Exception {
    final String json = new Gson().toJson(listOfItems());
    server.enqueue(new MockResponse().setBody(json));

    MyService sut = new MyServiceImpl(server.url("/").toString());
    sut.getItems(new RestCallback<List<Item>>() {
      @Override
      public void onResponse(List<Item> response) {
        L.INSTANCE.d(TAG, String.format("Received: %s", response.toString()));
        for (Item i : response) {
          assertTrue(json.contains(i.getDescription()));
          assertTrue(json.contains(i.getId()));
        }
      }

      @Override
      public void onFailure(String error) {
        L.INSTANCE.e(TAG, error);
        fail();
      }
    });
  }

  private class Callback<T> implements RestCallback<T> {
    private Predicate<T> predicate;

    Callback(Predicate<T> predicate) {
      this.predicate = predicate;
    }

    @Override
    public void onResponse(T response) {
      L.INSTANCE.d(TAG, String.format("Received: %s", response));
      assertTrue(String.format("%s matches with %s", predicate.toString(), response),
          predicate.apply(response));
    }

    @Override
    public void onFailure(String error) {
      L.INSTANCE.e(TAG, String.format("Error: %s", error));
    }
  }

  private List<Item> listOfItems() {
    List<Item> generated = new ArrayList<>();
    for (int i = 0; i < 10; i++) {
      generated.add(new Item(String.format("%d", System.nanoTime() + i), rndStr(10)));
    }
    return generated;
  }

  private String rndStr(int length) {
    return new RandomString(length, ThreadLocalRandom.current()).nextString();
  }
}

File: app/src/androidTest/java/com/amyllykoski/asyncseq/MyServiceImplTest.kt
package com.amyllykoski.asyncseq

import arrow.core.Either
import com.amyllykoski.asyncseq.api.MyError
import com.amyllykoski.asyncseq.impl.MyServiceImpl
import com.amyllykoski.asyncseq.model.Item
import com.amyllykoski.asyncseq.util.RandomString
import com.google.gson.Gson
import junit.framework.Assert.*
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.junit.Before
import org.junit.Test
import java.util.*
import java.util.concurrent.ThreadLocalRandom

class MyServiceImplTest {

  private lateinit var server: MockWebServer

  @Before
  @Throws(Exception::class)
  fun setUp() {
    server = MockWebServer()
    server.start()
  }

  @Test
  @Throws(Exception::class)
  fun getItemsFromMockServer() {
    val json = Gson().toJson(listOfItems())
    server.enqueue(MockResponse().setResponseCode(200).setBody(json))

    val sut = MyServiceImpl(server.url("/").toString())
    val response = sut.getItems()
    response.map {
      // List<Item>
      it.map {
        // Item
        assertTrue(json.contains(it.description))
        assertTrue(json.contains(it.id))
      }
    }
  }

  @Test
  @Throws(Exception::class)
  fun getItemsWithError() {
    server.enqueue(MockResponse().setResponseCode(500))
    val sut = MyServiceImpl(server.url("/").toString())

    // Should be a Left value.
    val response = sut.getItems()
    assertTrue(response.isLeft())

    // The error type should be ServerError.
    fun shouldBeServerError(response: Either<MyError, List<Item>>) {
      when (response) {
        is Either.Left -> when (response.a) {
          !is MyError.ServerError -> fail()
        }
        else -> fail()
      }
    }

    // Left value is not mapped.
    fun shouldNotMap() {
      response.map {
        it.map {
          fail()
        }
      }
    }

    shouldBeServerError(response)
    shouldNotMap()
  }

  private fun listOfItems(): List<Item> {
    val generated = ArrayList<Item>()
    for (i in 0 until NUM_OF_ITEMS) {
      generated.add(Item(String.format("%d", System.nanoTime() + i), rndStr(10)))
    }
    return generated
  }

  private fun rndStr(length: Int): String {
    return RandomString(length, ThreadLocalRandom.current()).nextString()
  }

  companion object {
    private const val NUM_OF_ITEMS = 10
  }
}